### PR TITLE
[WIP] Enable reuse of `ViewerModel` (close and show again)

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR (still WIP) will allow us to reuse a ViewerModel.  Currently, we just bail and say that if you've closed the viewer, you have to create a new one... but the window should be an on-demand view of a `ViewerModel`.   Mostly this means making sure that all of the connections between the `ViewerModel` and the "old" window are pruned when the window is destroyed (because the `ViewerModel.events` object outlives the Qt objects to which they are connect).    working example:

https://user-images.githubusercontent.com/1609449/129640782-b78654ed-a3c7-4834-b5bd-d5dc0540afe9.mov

at the moment, just trying to get it all to work, but will cleanup the code when it's working.

closes #2238 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Review by alisterburt — COMMENTED** [review]
oooh this is exciting! Only had a quick look but it's not jumping out at me, will custom viewer callbacks be preserved?

**Comment by tlambert03** [comment]
> oooh this is exciting! Only had a quick look but it's not jumping out at me, will custom viewer callbacks be preserved?

That would be the intention.  Currently I'm just trying to prune connections to qt objects... so it should leave pure python ones intact. But need to test

**Comment by sofroniewn** [comment]
@tlambert03 any further thoughts on this one? I think it would be a great one to get over the finish line.

I was looking over the implementation. One thing we could try and do is to "register" callbacks that we add internally in our objects (like vispylayer objects) and then call an explicit connect and disconnect. We definitely don't have to take that approach, but it could be nice. Anyway, just a thought. I'd be happy to pair on this one/ discuss it if you want

**Comment by brisvag** [comment]
I just tested some code on this branch, and I can confirm that this PR closes #3184.

**Comment by tlambert03** [comment]
tests are partially failing because of the recent qtconsole release: https://github.com/jupyter/qtconsole/issues/507

**Comment by psobolewskiPhD** [comment]
I love the idea of this, because I'm always been confused about how closing et al works with napari. If I run napari from the terminal it's clear, when i close it poofs with the python process too. But if I launch from ipython/VS Code close the window the process (with napari logo) persists. 
Also, on Mac there has always been a clear distinction between close and quit.

Curious how will actual quitting work?
File menu now has `File > Close window`, which currently acts more like Quit. So that will be become `File > Quit` and the new close feature will take over `File > Close window` ?

**Comment by tlambert03** [comment]
>  I'm always been confused about how closing et al works with napari.

I'm not sure any of this PR is going to really address this :).  This is just making sure that we can reuse a viewer model.  that is, create a new window for a pre-existing `ViewerModel` without worrying about events that were connected to the previous QMainWindow and children...

> But if I launch from ipython/VS Code close the window the process (with napari logo) persists.

This isn't going to change that. That happens because of these two lines:
https://github.com/napari/napari/blob/20211b7bb06055efa9b4368a9a8067ec47d9c3aa/napari/_qt/qt_event_loop.py#L168-L169

there one (and only one) [QApplication](https://doc.qt.io/qt-5/qapplication.html) per python process (regardless of how many viewers, or even if you use some other Qt-using thing like matplotlib).  And if you start napari within that process, and no logo has been assigned, we assign one

> Also, on Mac there has always been a clear distinction between close and quit.

basically, just think of a viewer as a window, and quit as "the application".  Many applications (including napari) have the rule that if you close the last top level window, you also quit the app.  It's the same here.  If we used multiple viewers more often, then closing one viewer would not quit the program.

**Comment by sofroniewn** [comment]
Any thoughts on picking this up again @tlambert03 - maybe @Carreau could help get it over the finish line and merged too?

**Comment by tlambert03** [comment]
> Any thoughts on picking this up again @tlambert03 - maybe @Carreau could help get it over the finish line and merged too?

Yeah, would be great.  I don't think this is a quick bump and over the finish line though.  Our qt objects are just too interconnected, and QtViewer touches a lot of stuff...

I'm going to close this, since at this point it probably needs a restart.  For myself, or anyone who wants to pick this up, a good starting point would be to add the [`test_reuse.py` file here](https://github.com/napari/napari/blob/5918a80019e917824d2ab649aab0fcd73a869875/napari/_tests/test_reuse.py), and try to make it pass.

adding here as well in case that link fails in the future:

```python
import numpy as np
from vispy.app.canvas import MouseEvent

import napari


def test_closed_viewer_ok(make_napari_viewer):
    v: napari.Viewer = make_napari_viewer(show=True)
    v.close()
    v.add_points()
    v.add_image(np.random.rand(4, 4))
    v.show()
    v.window.qt_viewer.on_mouse_move(MouseEvent('move'))
    v._on_cursor_position_change(None)
    v.close()
    v.show()


def test_reuse():
    v = napari.Viewer()
    v.add_points()
    v.add_image(np.random.rand(4, 4))
    v.close()
    v.show()


def test_add_then_show():
    v = napari.Viewer(show=False)
    v.add_points()
    v.add_image(np.random.rand(4, 4))
    v.show()


def test_qt_disconnected(make_napari_viewer):
    v = napari.Viewer()
    v.add_points()
    v.add_image(np.random.rand(8, 8))
    v.add_labels(np.random.rand(8, 8).astype(int))
    v.close()

```

the general goal of this PR was to make `viewer.window` a thing that can be `None`... and when `viewer.show()` is called, it will create a new window & view as needed, and when `viewer.close()` is called, the view and all Qt objects are cleaned up
